### DISCUSSION TITLE
Azure App Insights broken when using ServiceDefaults

### DISCUSSION THREAD

**Discussion by natelaff** [microsoft/aspire#4687] [Q&A] [answered]
Well #4322 was closed, but I still have an issue with app insights not pushing in my app when running on Azure.

I added Orchestrator support to my app and all seemed great. But when running on Azure App Services, I noticed app insights were no longer pushing.

```
var builder = WebApplication.CreateBuilder(args);
builder.AddServiceDefaults();

var config = builder.Configuration;

builder.Services.AddApplicationInsightsTelemetry(options => {
    options.DeveloperMode = builder.Environment.IsDevelopment();
    options.ConnectionString = config["APPLICATIONINSIGHTS_CONNECTION_STRING"];
```

The issue seems to stem from the call to

```
builder.ConfigureOpenTelemetry();
```

in the ServiceDefaults. If i comment that out and run in Azure, my insights start working again.

**Comment by davidfowl** [comment]
cc @samsp-msft

**Comment by samsp-msft** [comment, marked as answer]
It looks like you are mixing the classic Application Insights SDK and OpenTelemetry. Aspire is built using OpenTelemetry and the OTel based exporters to AzureMonitor/Application Insights. The call to ServiceDefaults is setting up OpenTelemetry. 

To use Application Insights with Aspire, please use either Azure Monitor Distro or Azure monitor exporter. The former is a wrapper around the latter that provides good defaults and extensions to provide the best experience.

https://learn.microsoft.com/en-us/azure/azure-monitor/app/opentelemetry-enable?tabs=aspnetcore#enable-azure-monitor-opentelemetry-for-net-nodejs-python-and-java-applications
https://learn.microsoft.com/en-us/dotnet/aspire/deployment/azure/application-insights

**Comment by natelaff** [comment]
Ok, I will look into those thanks. The issue in my opinion here is adding App Insights telemetry to the project via right click gives you this code. Adding Aspire orchestration which then adds the Service defaults gives you conflicting code. Something to consider ;)

**Reply by davidfowl** [reply]
@samsp-msft This is worth a documentation update.

**Comment by DGAISmith** [comment]
Is there a use case in this scenario where you would want to use AddApplicationInsightsTelemetry when deployed to azure, but ConfigureOpenTelemetry when running locally.  Thinking mainly about when existing services which already use AppInsights have Aspire support added to them, and a developer might want to use OTEL and the Aspire Dashboard when running locally, but retain the existing AppInsights when deployed (via whatever existing deployment mechanism the service already has in place)

**Reply by davidfowl** [reply]
Absolutely! Though there are several ways to do that. OpenTelemetry has a concept of exporters and there's one for azure monitor (that replaces calls to AddApplicationInsightsTelemetry). So you're funneling everything through a single system.

If you can't upgrade to the new APIs for some reason, it should be possible to make them work side by side or only in dev and only when deployed.

**Comment by boclifton-MSFT** [comment]
It isn't entirely clear to me from the documentation that I've seen how to consume or utilize the App Insights exporter in the client apps.  Can I just start seeing telemetry in App Insights without any further configuration than the Service Defaults project, or do I need to follow the Aspire convention of adding with a `builder.AddAppInsights()` call in Program.cs?  The docs above show how to add it to the AppHost project to programmatically deploy the resource and how to configure Service Defaults to ensure it's configured properly.  But does the Aspire integration have a Nuget package for the consumers to get a preconfigured TelemetryClient or similar?  If not or if this is just entirely the wrong question, it may be worth adding something to the documentation for those of us a little unfamiliar with how OpenTelemetry works with Aspire.

**Reply by davidfowl** [reply]
Are you talking about this document https://learn.microsoft.com/en-us/dotnet/aspire/deployment/azure/application-insights ?

cc @samsp-msft @IEvangelist

**Reply by boclifton-MSFT** [reply]
Yeah, that doc eventually gets me to a working App Insights setup.  But there's not really any mention of using App Insight's TelemetryClient or any other client-based implementation for custom telemetry.  Is that supported?  If so, how do I do that in an Aspire app?

**Reply by davidfowl** [reply]
Got it. Perhaps we should link to the official azure monitor documentation here:

<img width="337" alt="image" src="https://github.com/user-attachments/assets/a7af8f50-3f96-4c44-a924-177b4c73bd0f">

Notice the TOC has 2 entries, the "classic SDK", this is what people have typically used with AppInsights. Aspire recommends using the Otel distro instead.

Start from https://learn.microsoft.com/en-us/azure/azure-monitor/app/opentelemetry-add-modify?tabs=aspnetcore.
